### PR TITLE
drivers:platform:pico:pico_spi: Add support for multiple slaves

### PR DESCRIPTION
**Pull request by rbolboac**
Add support for spi driver on pico platform such that communication is supported with multiple spi slaves, with different bus configuration.

### CODE REVIEW

File: drivers/platform/pico/pico_spi.c [hunk 1]
```diff
@@ -52,62 +52,36 @@
...
- */
-int32_t pico_spi_init(struct no_os_spi_desc **desc,
-		      const struct no_os_spi_init_param *param)
+static int pico_spi_config(struct no_os_spi_desc *desc)
 {
-	int32_t ret;
-	struct no_os_spi_desc *descriptor;
 	struct pico_spi_desc *pico_spi;
-	struct pico_spi_init_param *pico_spi_ip;
+	pico_spi = desc->extra;
```
Comment (CiprianRegus) [inline drivers/platform/pico/pico_spi.c:58]: Maybe assign it directly on the previous line, or below the variable declaration block.

File: drivers/platform/pico/pico_spi.c [hunk 2]
```diff
@@ -200,6 +221,18 @@ int32_t pico_spi_write_and_read(struct no_os_spi_desc *desc,
 
 	pico_spi = desc->extra;
 
+	// Compute a slave ID based on SPI instance and chip select.
+	// If it did not change since last call to pico_spi_write_and_read,
+	// no need to reconfigure SPI. Otherwise, reconfigure it.
```
Comment (CiprianRegus) [inline drivers/platform/pico/pico_spi.c:226]: nit: `/* */`

File: drivers/platform/pico/pico_spi.c [hunk 3]
```diff
@@ -125,27 +99,71 @@ int32_t pico_spi_init(struct no_os_spi_desc **desc,
...
+
+
+	if (!desc || !param || !param->extra)
+		return -EINVAL;
+
+	descriptor = (struct no_os_spi_desc *)calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	pico_spi = (struct pico_spi_desc *) calloc(1, sizeof(*pico_spi));
```
Comment (CiprianRegus) [inline drivers/platform/pico/pico_spi.c:141]: Extra space after the cast.

File: drivers/platform/pico/pico_spi.c [hunk 4]
```diff
@@ -125,27 +99,71 @@ int32_t pico_spi_init(struct no_os_spi_desc **desc,
...
+ */
+int32_t pico_spi_init(struct no_os_spi_desc **desc,
+		      const struct no_os_spi_init_param *param)
+{
+	int32_t ret;
+	struct no_os_spi_desc *descriptor;
+	struct pico_spi_desc *pico_spi;
+	struct pico_spi_init_param *pico_spi_ip;
+
+
```
Comment (CiprianRegus) [inline drivers/platform/pico/pico_spi.c:133]: Extra line

**Review by CiprianRegus — COMMENTED** [review]
Looks good, styling only.

**Comment by rbolboac** [comment]
- implemented review changes

**Review by CiprianRegus — APPROVED** [review]

File: drivers/platform/pico/pico_spi.c [hunk 5]
```diff
@@ -125,27 +100,70 @@ int32_t pico_spi_init(struct no_os_spi_desc **desc,
...
+		return -EINVAL;
+
+	descriptor = (struct no_os_spi_desc *)calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	pico_spi = (struct pico_spi_desc *)calloc(1, sizeof(*pico_spi));
+	if (!pico_spi) {
+		ret = -ENOMEM;
+		goto error;
```
Comment (amiclaus) [inline drivers/platform/pico/pico_spi.c:144]: should we free `pico_spi` in this case?

Comment (rbolboac) [reply drivers/platform/pico/pico_spi.c]: freeing a null pointer will not generate an error, however I added a new label to free only the descriptor in this case.

**Pull request merged**